openpype/hosts/max/plugins/publish/validate_loaded_plugin.py

```diff
@@ -0,0 +1,73 @@
+# -*- coding: utf-8 -*-
+"""Validator for USD plugin."""
+from pyblish.api import InstancePlugin, ValidatorOrder
+from pymxs import runtime as rt
+
+from openpype.pipeline.publish import (
+    RepairAction,
+    OptionalPyblishPluginMixin,
+    PublishValidationError
+)
+from openpype.hosts.max.api.lib import get_plugins
+
+
+class ValidateLoadedPlugin(OptionalPyblishPluginMixin,
+                           InstancePlugin):
+    """Validates if the specific plugin is loaded in 3ds max.
+    User can add the plugins they want to check through"""
+
+    order = ValidatorOrder
+    hosts = ["max"]
+    label = "Validate Loaded Plugin"
+    optional = True
+    actions = [RepairAction]
+
+    def get_invalid(self, instance):
+        """Plugin entry point."""
+        if not self.is_active(instance.data):
+            self.log.debug("Skipping Validate Loaded Plugin...")
+            return
+        invalid = []
+        # display all DLL loaded plugins in Max
+        plugin_info = get_plugins()
+        project_settings = instance.context.data[
+            "project_settings"]["max"]["publish"]
+        target_plugins = project_settings[
+            "ValidateLoadedPlugin"]["plugins_for_check"]
+        for plugin in target_plugins:
+            if plugin.lower() not in plugin_info:
+                invalid.append(
+                    f"Plugin {plugin} not exists in 3dsMax Plugin List.")
+            for i, _ in enumerate(plugin_info):
+                if plugin.lower() == rt.pluginManager.pluginDllName(i):
+                    if not rt.pluginManager.isPluginDllLoaded(i):
+                        invalid.append(
+                            f"Plugin {plugin} not loaded.")
+        return invalid
+
+    def process(self, instance):
+        invalid_plugins = self.get_invalid(instance)
+        if invalid_plugins:
+            bullet_point_invalid_statement = "\n".join(
+                "- {}".format(invalid) for invalid in invalid_plugins
+            )
+            report = (
+                "Required plugins fails to load.\n\n"
+                f"{bullet_point_invalid_statement}\n\n"
+                "You can use repair action to load the plugin."
+            )
+            raise PublishValidationError(
+                report, title="Required Plugins unloaded")
+
+    @classmethod
+    def repair(cls, instance):
+        plugin_info = get_plugins()
+        project_settings = instance.context.data[
+            "project_settings"]["max"]["publish"]
+        target_plugins = project_settings[
+            "ValidateLoadedPlugin"]["plugins_for_check"]
+        for plugin in target_plugins:
+            for i, _ in enumerate(plugin_info):
+                if plugin == rt.pluginManager.pluginDllName(i):
+                    if not rt.pluginManager.isPluginDllLoaded(i):
+                        rt.pluginManager.loadPluginDll(i)
```

Comment:
It doesn't make much sense that this is an `InstancePlugin`. The plug-ins are global to your scene and thus this check should only really run once.

Also, I'd first check if there are **any** plugins to validate in the settings before retrieving anything from 3DsMax - so to the `get_plugins` call later. At least that's what I figured from the docstring - to check if a required plugin is loaded. However, the validator seems to actually validate whether only **those** plugins are loaded and **no others** are loaded. So it's quite strict - is that how it's intended? (Note that the `if not target_plugins` in my suggestion changes that a bit)

Here's a quick refactor to ContextPlugin. Note you might need to update imports:

`InstancePlugin` -> `ContextPlugin`
`RepairAction` -> `RepairContextAction`

```suggestion
class ValidateLoadedPlugin(OptionalPyblishPluginMixin,
                           ContextPlugin):
    """Validates if the specific plugin is loaded in 3ds max.
    Studio settings defines the plugins to validate."""

    order = ValidatorOrder
    hosts = ["max"]
    label = "Validate Loaded Plugins"
    optional = True
    actions = [RepairContextAction]

    def get_invalid(self, context):
        """Plugin entry point."""
        if not self.is_active(context.data):
            self.log.debug("Skipping Validate Loaded Plugin...")
            return
        project_settings = context.data[
            "project_settings"]["max"]["publish"]
        target_plugins = project_settings[
            "ValidateLoadedPlugin"]["plugins_for_check"]
        if not target_plugins:
            return
            
        # get all DLL loaded plugins in Max
        plugin_info = get_plugins()
        invalid = []
        for plugin in target_plugins:
            if plugin.lower() not in plugin_info:
                invalid.append(
                    f"Plugin {plugin} does not exist in 3dsMax Plugin List.")
            for i, _ in enumerate(plugin_info):
                if plugin.lower() == rt.pluginManager.pluginDllName(i):
                    if not rt.pluginManager.isPluginDllLoaded(i):
                        invalid.append(
                            f"Plugin {plugin} not loaded.")
        return invalid

    def process(self, context):
        invalid_plugins = self.get_invalid(context)
        if invalid_plugins:
            bullet_point_invalid_statement = "\n".join(
                "- {}".format(invalid) for invalid in invalid_plugins
            )
            report = (
                "Required plugins fails to load.\n\n"
                f"{bullet_point_invalid_statement}\n\n"
                "You can use repair action to load the plugin."
            )
            raise PublishValidationError(
                report, title="Required Plugins unloaded")

    @classmethod
    def repair(cls, context):
        plugin_info = get_plugins()
        project_settings = context.data[
            "project_settings"]["max"]["publish"]
        target_plugins = project_settings[
            "ValidateLoadedPlugin"]["plugins_for_check"]
        for plugin in target_plugins:
            for i, _ in enumerate(plugin_info):
                if plugin == rt.pluginManager.pluginDllName(i):
                    if not rt.pluginManager.isPluginDllLoaded(i):
                        rt.pluginManager.loadPluginDll(i)
```

Reply:
Also, this might make for a more readable (and optimized version):
```suggestion
class ValidateLoadedPlugin(OptionalPyblishPluginMixin,
                           ContextPlugin):
    """Validates if the specific plugin is loaded in 3ds max.
    User can add the plugins they want to check through"""

    order = ValidatorOrder
    hosts = ["max"]
    label = "Validate Loaded Plugins"
    optional = True
    actions = [RepairContextAction]

    def get_invalid(self, context):
        """Plugin entry point."""
        if not self.is_active(context.data):
            self.log.debug("Skipping Validate Loaded Plugin...")
            return
        invalid = []
        
        # get all DLL loaded plugins in Max and their plugin index
        available_plugins = {
            plugin_name.lower(): index for index, plugin_name in get_plugins()
        }
        required_plugins = (
            context.data["project_settings"]["max"]["publish"]
                        ["ValidateLoadedPlugin"]["plugins_for_check"]
        )
        for plugin in required_plugins:
            plugin_name = plugin.lower()
            
            plugin_index = available_plugins.get(plugin_name)
            if plugin_index is None:
                invalid.append(
                    f"Plugin {plugin} not exists in 3dsMax Plugin List."
                )
                continue
                
            if not rt.pluginManager.isPluginDllLoaded(plugin_index):
                invalid.append(f"Plugin {plugin} not loaded.")
                
        return invalid

    def process(self, context):
        invalid_plugins = self.get_invalid(context)
        if invalid_plugins:
            bullet_point_invalid_statement = "\n".join(
                "- {}".format(invalid) for invalid in invalid_plugins
            )
            report = (
                "Required plugins fails to load.\n\n"
                f"{bullet_point_invalid_statement}\n\n"
                "You can use repair action to load the plugin."
            )
            raise PublishValidationError(
                report, title="Required Plugins unloaded")

    @classmethod
    def repair(cls, context):
        plugin_info = get_plugins()
        project_settings = context.data[
            "project_settings"]["max"]["publish"]
        target_plugins = project_settings[
            "ValidateLoadedPlugin"]["plugins_for_check"]
        for plugin in target_plugins:
            for i, _ in enumerate(plugin_info):
                if plugin == rt.pluginManager.pluginDllName(i):
                    if not rt.pluginManager.isPluginDllLoaded(i):
                        rt.pluginManager.loadPluginDll(i)
```